Simplify Product Preferences Form

Progi1984 opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplify Product Preferences Form
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  | Simplifying Configuration -> Shop Parameters -> Product Preferences. Everything must look the same aside for help now appearing under inputs instead as blue box, also some fields now will appear as required because they always were.

:notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType by PageType and PaginationType. This means if any module extends those types they will get an exception upon upgrading to PS version containing changes in this PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/27892)
<!-- Reviewable:end -->

Progi1984 committed:
Simplify Product Preferences Form

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
+                        'Catalog mode disables the shopping cart on your store. Visitors will be able to browse your products catalog, but not buy them.',
+                        'Admin.Shopparameters.Help'
+                    ) . '<br />' . $this->trans(
+                        'Have specific needs? Edit particular groups to let them see prices or not.',
```
This help text could be clearer on where to perform the action.

```suggestion
                        'To hide prices for a specific group, go to [1]Customer Settings > Groups[\1].',                        
```

Also, this help text should be placed under the "Show prices" field instead of "Catalog mode".

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Set the maximum size of the summary of your product description (in characters).',
```
I think we can remove this tooltip. The label is already clear enough and it's repetitive. + It's not clear: are we talking about the product summary or the description?

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
+            ])
+            ->add('default_status', SwitchType::class, [
+                'label' => $this->trans(
+                    'Default activation status',
```
```suggestion
                    'Activate new products by default',
```

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/GeneralType.php:
```diff
@@ -46,25 +46,78 @@ class GeneralType extends TranslatorAwareType
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'When active, new products will be activated by default during creation.',
```
This help text can be removed.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -37,27 +37,84 @@
+            ])
+            ->add('display_last_quantities', IntegerType::class, [
+                'label' => $this->trans(
+                    'Display remaining quantities when the quantity is lower than',
```
I tested that feature but couldn't make it work. Where are the remaining quantities supposed to appear? 🤔

Progi1984 replied:
It impacts the configuration `PS_LAST_QTIES`. That impacts in front the label `Last items in stock`.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -37,27 +37,84 @@
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'In the volume discounts board, display the new price with the applied discount instead of showing the discount (ie. "-5%").',
```
What "discounted prices" are we talking about? Specific prices? Because I've just tested the feature and couldn't see the difference between when it's enabled and when it's disabled. 🤔 I don't understand what is the "volume discounts board".

Progi1984 replied:
That impacts the configuration `PS_DISPLAY_DISCOUNT_PRICE`. It used with Specific Prices. And in front, if enabled, that shows the price else the discount. I think you can see it in volumes discounts table on the front office.

![image](https://user-images.githubusercontent.com/1533248/158846381-781f7727-fc34-4f58-9c3e-4a31ba4002dd.png)

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -36,16 +36,33 @@
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Number of products displayed per page. Default is 10.',
```
This help text can be removed, the label speaks for itself.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PaginationType.php:
```diff
@@ -59,6 +76,14 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'Default order method for product list.',
```
This help text can be removed, the label speaks for itself.

Julievrz reviewed (commented):
A few suggestions and questions about wording

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -37,27 +37,84 @@
+                    'Admin.Shopparameters.Feature'
+                ),
+                'help' => $this->trans(
+                    'In the volume discounts board, display the new price with the applied discount instead of showing the discount (ie. "-5%").',
```
```suggestion
                    'In the volume discount table on the product page, display the discounted unit price instead of the unit discount. E.g. If you sell a product for $10 with a discount of $2 from 3 items purchased, the discounted unit price ($8) will be displayed instead of the unit discount ($2).',
```

Progi1984 replied:
Fixed

Julievrz commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/PageType.php:
```diff
@@ -37,27 +37,84 @@
-            ->add('display_discount_price', SwitchType::class);
+            ->add('display_discount_price', SwitchType::class, [
+                'label' => $this->trans(
+                    'Display discounted price',
```
```suggestion
                    'Display the discounted unit price',
```

Progi1984 replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml:
```diff
@@ -1839,17 +1855,3 @@ services:
     tags:
       - { name: form.type }
-
-  form.type.common.zone_choice:
```
Why remove those forms? Is it maybe a typo or a mistake after a rebase (since I believe you worked on these types recently in another PR maybe some contents were mixed)?

It doesn't seem related to the PR anyway so it seems strange.

jolelievre reviewed (commented):
Thanks @Progi1984 

I just want to confirm with you that the removal of `ZoneChoiceType` and `ConfigurableCountryChoiceType` service definitions is on purpose because it doesn't seem related to the rest of this PR as they are not even used by the improved product form types.

jolelievre commented:
@matks or @matthieu-rolland can you pleas re-approve this PR?

marwachelly reviewed (commented):
Hello @Progi1984 ,

I tested the PR and I noticed the following issues:

1. With PR some field are marked as required. Is there any spec related?

![image](https://user-images.githubusercontent.com/89604350/163977981-2da1177b-02cb-4fdf-9d7b-a7a1407ae90b.png)


2. Some field are underlined and lead to the page Shop parameters> Customer settings > Groups

https://user-images.githubusercontent.com/89604350/164002783-930f13a2-6147-4018-827b-84bef0dd5c22.mp4




3. Some help text are missing. Not all blue box are replaced.
![image](https://user-images.githubusercontent.com/89604350/163994956-ece20f2b-40f1-4e96-b459-01b85194820f.png)

4. The behavior of the filed "show prices" is changed. According to the spec "It should be displayed only when the catalog mode is enabled"
Before:

https://user-images.githubusercontent.com/89604350/164003176-f1068f08-0839-4f93-8ce2-6f015cc8e377.mp4



With PR:


https://user-images.githubusercontent.com/89604350/164002429-b053c0a1-7e22-4e2b-b961-a1b225aa10a0.mp4



5. The order of fields in "products stock" section is changed
Without PR 
![image](https://user-images.githubusercontent.com/89604350/164003507-b4e31922-46db-43bf-bbb0-b05dedd383b0.png)


with PR
![image](https://user-images.githubusercontent.com/89604350/163999803-1ac3793c-bd29-42c4-a9e4-1292c941b3f3.png)

6. A toggle button without field in the section "products stock" 
![image](https://user-images.githubusercontent.com/89604350/164000210-39044da9-40df-462c-8e52-6b13eb2ae296.png)




7. the drop down of the field "Default order of priority for specific prices " is changed
without PR:
![image](https://user-images.githubusercontent.com/89604350/163997149-e5d4807a-07d8-475f-a41b-063ca62f0f11.png)

With PR:
![image](https://user-images.githubusercontent.com/89604350/163997215-f78a79d9-aee4-4b83-97f7-f3357f53e93c.png)

8. The help text and the name of the field "Display discounted price" are changed
Without PR
![image](https://user-images.githubusercontent.com/89604350/163997482-14a4a0d6-ca13-483e-a0e7-5c12af3abe93.png)
With PR
![image](https://user-images.githubusercontent.com/89604350/163997534-d93d6010-84cf-4d10-b0af-67c8876c7b5e.png)

Could you please check? 

Thanks,

Progi1984 commented:
@marwachelly 

1. Fixed (I remove all required fields)
2. Fixed (I have bad closed a `<a>` tag)
3. @Julievrz asked to remove some help texts
  * 1 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827795130
  * 2 : Fixed
  * 3 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827816261
  * 4 : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r827816429
4. Fixed
5. Fixed
6. Fixed
7. @NeOMakinG Is it the new style ? 
Old style : ![image](https://user-images.githubusercontent.com/89604350/163997149-e5d4807a-07d8-475f-a41b-063ca62f0f11.png)
New style : ![image](https://user-images.githubusercontent.com/89604350/163997215-f78a79d9-aee4-4b83-97f7-f3357f53e93c.png)
9. @Julievrz asked to change the help text : https://github.com/PrestaShop/PrestaShop/pull/27892#discussion_r833452867

Progi1984 commented:
@PierreRambaud & @jolelievre Fixed some problems for @marwachelly

atomiix commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -47,59 +47,151 @@ class StockType extends TranslatorAwareType
+                    'Admin.Shopparameters.Help'
+                ),
+                'choices' => [
+                    'Decrement pack only.' => 0,
+                    'Decrement products in pack only.' => 1,
+                    'Decrement both.' => 2,
```
I think these have to be translated

Progi1984 replied:
The translation is done automatically.

atomiix replied:
How? 🤔 Using which domain?

Progi1984 replied:
Fixed. Check the key `choice_translation_domain`.

Progi1984 commented:
@matks & @atomiix Fixed feedbacks from @atomiix

jolelievre commented on src/PrestaShopBundle/Form/Admin/Configure/ShopParameters/ProductPreferences/StockType.php:
```diff
@@ -47,59 +47,152 @@ class StockType extends TranslatorAwareType
+                    'Admin.Shopparameters.Help'
+                ),
+                'choices' => [
+                    'Decrement pack only.' => 0,
```
I think you should use const here `PrestaShop\PrestaShop\Core\Domain\Product\Pack\ValueObject\PackStockType::STOCK_TYPE_PACK_ONLY`

Progi1984 replied:
Fixed

jolelievre reviewed (commented):
Sorry it was already approved, it's not blocking but I have a suggestion

Progi1984 commented:
@matks @jolelievre @atomiix Fixed feedbacks from @jolelievre

sLorenzini reviewed (commented):
hello @Progi1984,

@khouloudbelguith  and I have tested this PR, and we noticed the following issues:

All selected list on the page have a blank first option :

- Quantity discounts based on
- Separator of attribute anchor on the product links 
- Default pack stock management 
- Default order by 
- Default order method 

Without PR:

![selected-list-without-pr](https://user-images.githubusercontent.com/2854834/165104375-be7ba30e-daa7-482b-a59b-7707e868656e.png)

With PR:

![selected-list](https://user-images.githubusercontent.com/2854834/165102792-7363e26a-dbdc-44c3-aa2b-8d48c54faf13.png)

------------------------------

Issue on number of characters on "Max size of short description" field in product page v2

Change the Max size of short description to 20 and save on product settings page.
Go to BO > Catalog > Products V2 > Edit a product
See error > in the summary section, the max size is not updated

![max-size-characters](https://user-images.githubusercontent.com/2854834/165106006-10ef1c9c-b0aa-45eb-8d83-a8dcfabcd5c9.png)

Set the summary to more than 20 characters and save

![max-size-exception](https://user-images.githubusercontent.com/2854834/165106245-b4eda9cf-5f12-412c-8043-ef9ca6640605.png)

------------------------------

Issue  on "Force update of friendly URL" field in product page v2

Set to yes "Force update of friendly URL" switch button on product settings page.
Go to BO > Catalog > Products V2 > Edit a product
Go on "SEO" tab
Update the name of the product
See error > the "Friendly URL" field is not updated 

With product page V1:

https://user-images.githubusercontent.com/2854834/165105843-e001710c-a820-4744-ba35-7f1b9d557738.mp4

With product page V2:

https://user-images.githubusercontent.com/2854834/165105924-55694d5a-3513-4b7b-8df2-6b0d0251d95f.mp4

------------------------------

Issue  on "Default activation status" field in product page v2

Set to yes "Default activation status" switch button on product settings page.
Go to BO > Catalog > Products V2 > add a product
See error at the bottom > the "Offline" switch button is not active

With product page V1:

https://user-images.githubusercontent.com/2854834/165107223-7682c1f7-f89b-41d9-a1ca-ca024ee05490.mp4

With product page V2:

https://user-images.githubusercontent.com/2854834/165107259-5ec3db16-6756-4df3-9847-2dede75b24ad.mp4


------------------------------

There is a space between the "Default order of priority for specific prices" field and its help text:

![space-between-field-and-help-text](https://user-images.githubusercontent.com/2854834/165103651-c8624623-e0e9-4f6a-bd1e-d2d7176ac1df.png)

------------------------------

We have an issue when we set those fields with too long characters:
- Label of in-stock products
- Label of out-of-stock products with allowed backorders
- Label of out-of-stock products with denied backorders
- Delivery time of in-stock products
- Delivery time of out-of-stock products with allowed backorders

![image](https://user-images.githubusercontent.com/2854834/165109510-e7f8347b-0b59-44df-901f-09fb5bbb50ee.png)

![image (1)](https://user-images.githubusercontent.com/2854834/165109552-97ba17f1-2f0b-4e7f-8753-df3d5bfcf642.png)

![image (2)](https://user-images.githubusercontent.com/2854834/165109574-47ac7b66-511e-44ff-9396-103f227ccb73.png)

![image (3)](https://user-images.githubusercontent.com/2854834/165109987-4efc6bb5-55a2-4d62-ac7a-6623c71d11cb.png)

------------------------------

We have an issue on pagination block

Go to BO > Product settings page > Pagination Block
Set to 0 the "Products per page" field in product settings page
Go to FO > Home category
See error on top of the page

![error-pagination](https://user-images.githubusercontent.com/2854834/165108496-3ae09b70-c16b-4a9f-846a-87ccfb769214.png)

At the bottom of the page, the showing text must be hidden:

![showing-product-category-page](https://user-images.githubusercontent.com/2854834/165108699-019653c6-61c5-48d6-8cca-9f44e6b5e1d2.png)

Go to BO > Product settings page > Pagination Block
A help text could be added below the "Products per page" field  to inform that we can remove the pagination if it's set to 0

![product-per-page-set-0](https://user-images.githubusercontent.com/2854834/165109050-40391ca3-f401-4f9f-ba7e-e0bb08906c36.png)

------------------------------

When we set an invalid number in those fields for example --50

- Number of days for which the product is considered 'new'
- Display remaining quantities when the quantity is lower than
- Products per page

the alert is moving when we try to scroll

https://user-images.githubusercontent.com/2854834/165111416-4f220443-2cc1-47b1-be6e-4cfc398fb27a.mp4

------------------------------

We have an issue when we tested the product settings page in responsive mode on Mac OS **(with and without PR)**:

With responsive mode in chrome debugger:

![responsive-chrome-mac](https://user-images.githubusercontent.com/2854834/165110748-0f071a49-a059-49e1-8e83-26af0b1136a7.png)

With iphone mode in chrome debugger:

![iphone-chrome-mac](https://user-images.githubusercontent.com/2854834/165110720-2783f788-0954-4351-b7dc-5e5ba807c754.png)

With pixel mode in chrome debugger:

![pixel-chrome-mac](https://user-images.githubusercontent.com/2854834/165110680-bb03f60c-93d3-41c5-8a3f-48add29c4a52.png)


Tested on browserstack with iphone screen, linux and windows OS and it is good.

------------------------------

We have an issue on selected list in the product settings pages in Firefox on Mac OS **(with and without PR)**:

On some selected list, when we click on it to change the option, we cannot choose the option on the first click:

https://user-images.githubusercontent.com/2854834/165112658-8b47899c-44ea-4c8e-8c2f-1f28a546b979.mp4

Tested on browserstack on Firefox in windows and linus OS and it is good.


Thank you

Progi1984 commented:
@sLorenzini @khouloudbelguith My report.

* All selected list on the page have a blank first option :
  * Fixed
* Issue on number of characters on "Max size of short description" field in product page v2
  * Not in relation with this PR. Could you create an issue ?
* We have an issue when we set those fields with too long characters:
  * Not in relation with this PR. Could you create an issue ?
* We have an issue on pagination block 
  * Not in relation with this PR. Could you create an issue ?
* We have an issue when we tested the product settings page in responsive mode on Mac OS (with and without PR): 
  * Not in relation with this PR. Could you create an issue ?
* We have an issue on selected list in the product settings pages in Firefox on Mac OS (with and without PR): 
  * Not in relation with this PR. Could you create an issue ?

MeKeyCool reviewed (commented):
Sorry I don't feel comfortable enough to speak about specific symfony form refactoring or about behaviors you updated.

sLorenzini commented:
> @sLorenzini @khouloudbelguith My report.
> 
> * All selected list on the page have a blank first option :
>   
>   * Fixed
> * Issue on number of characters on "Max size of short description" field in product page v2
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue when we set those fields with too long characters:
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue on pagination block
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue when we tested the product settings page in responsive mode on Mac OS (with and without PR):
>   
>   * Not in relation with this PR. Could you create an issue ?
> * We have an issue on selected list in the product settings pages in Firefox on Mac OS (with and without PR):
>   
>   * Not in relation with this PR. Could you create an issue ?

@Progi1984 thanks for your report.

All issues not in relation with this PR have been created.

Thank you

khouloudbelguith reviewed (approved):
Hi @Progi1984,

It is ok :heavy_check_mark: 

https://user-images.githubusercontent.com/16067358/166726351-d857d210-74ef-4fa8-b040-37ce5d387fcd.mp4

https://user-images.githubusercontent.com/16067358/166726517-3829b4e0-f45c-4675-9658-412df129b88e.mp4


I have only a difference in the wording
**Without PR**
`Default activation status`
![image](https://user-images.githubusercontent.com/16067358/166726184-622cebc7-716e-4b53-991e-ca852004123a.png)
**With PR**
`Activate new products by default`
![image](https://user-images.githubusercontent.com/16067358/166726338-77fd15dc-905b-426a-bbde-4389b642f265.png)

Ping @Julievrz what do you think?
If it is ok, I validate this PR.

Thank you!

atomiix commented:
Thanks @Progi1984

Julievrz commented:
> Hi @Progi1984,
> 
> It is ok ✔️
> 
>  Part.2.mp4 
>  Part.1.mp4 
> I have only a difference in the wording **Without PR** `Default activation status` ![image](https://user-images.githubusercontent.com/16067358/166726184-622cebc7-716e-4b53-991e-ca852004123a.png) **With PR** `Activate new products by default` ![image](https://user-images.githubusercontent.com/16067358/166726338-77fd15dc-905b-426a-bbde-4389b642f265.png)
> 
> Ping @Julievrz what do you think? If it is ok, I validate this PR.
> 
> Thank you!

Too late 😅 It would have been better to remove the help text indeed. It's redundant, the label speaks for itself.

Julievrz commented:
We will do a final review of the wording of the product page anyway because it is really difficult to review as it's broken down into several pieces.
